python_modules/dagster/dagster_tests/core_tests/resource_tests/pythonic_resources/test_direct_invocation.py

```diff
@@ -475,6 +476,121 @@ async def main():
+        r" cannot be used to execute another op until .* has finished executing",
+    ):
+        asyncio.run(main())
+
+
+def test_direct_invocation_resource_context_manager():
+    from dagster import resource
+
+    class YieldedResource:
+        def get_value(self):
+            return 1
+
+    @resource
+    def yielding_resource(context):
+        yield YieldedResource()
+
+    @asset(required_resource_keys={"yielded_resource"})
+    def my_asset(context):
+        assert context.resources.yielded_resource.get_value() == 1
+
+    with build_op_context(resources={"yielded_resource": yielding_resource}) as ctx:
+        my_asset(ctx)
+
+
+def test_context_bound_state_non_generator():
+    @asset
+    def my_asset(context):
+        assert context._bound_properties is not None  # noqa: SLF001
+
+    ctx = build_op_context()
+    assert ctx._bound_properties is None  # noqa: SLF001
+
+    my_asset(ctx)
+    assert ctx._bound_properties is None  # noqa: SLF001
+    assert ctx._invocation_properties is not None  # noqa: SLF001
+
+
+def test_context_bound_state_generator():
+    @op(out={"first": Out(), "second": Out()})
+    def generator(context):
+        assert context._bound_properties is not None  # noqa: SLF001
+        yield Output("one", output_name="first")
+        yield Output("two", output_name="second")
+
+    ctx = build_op_context()
+
+    result = list(generator(ctx))
+    assert result[0].value == "one"
+    assert result[1].value == "two"
+
+    assert ctx._bound_properties is None  # noqa: SLF001
+    assert ctx._invocation_properties is not None  # noqa: SLF001
+
+
+def test_context_bound_state_async():
+    @asset
+    async def async_asset(context):
+        assert context._bound_properties is not None  # noqa: SLF001
+        assert context.asset_key.to_user_string() == "async_asset"
+        await asyncio.sleep(0.01)
+        return "one"
+
+    ctx = build_asset_context()
+
+    result = asyncio.run(async_asset(ctx))
+    assert result == "one"
+
+    assert ctx._bound_properties is None  # noqa: SLF001
+    assert ctx._invocation_properties is not None  # noqa: SLF001
+
+
+def test_context_bound_state_async_generator():
+    @op(out={"first": Out(), "second": Out()})
+    async def async_generator(context):
+        assert context._bound_properties is not None  # noqa: SLF001
+        yield Output("one", output_name="first")
+        await asyncio.sleep(0.01)
+        yield Output("two", output_name="second")
+
+    ctx = build_op_context()
+
+    async def get_results():
+        res = []
+        async for output in async_generator(ctx):
+            res.append(output)
+        return res
+
+    result = asyncio.run(get_results())
+    assert result[0].value == "one"
+    assert result[1].value == "two"
+
+    assert ctx._bound_properties is None  # noqa: SLF001
+    assert ctx._invocation_properties is not None  # noqa: SLF001
+
+
+def test_bound_state_with_error():
```

Comment:
i think this is the last behavior case we need to think through. Now that we call `unbind` at the end of execution, if an invocation fails midway through execution, the context is never unbound, which causes the next invocation to fail because the context is still bound. 

Options:
1. leave this behavior as is. If users want to check for an expected failure and then reuse the context they need to call `unbind` themselves
2. try to catch the thrown exceptions from the invocation and call `unbind`

Reply:
we'd have to shuffle the logic to get the `context` out of args/kwargs into the `__call__` function, but that should be doable. Not 100% sure this will handle asyncio cases though, but it's worth experimenting

Reply:
Ended up needing to do a try catch in a couple places:
1. in `direct_invocation_result` wrapping all code after the context was bound
2. in each of the asyn/generator/asyn generator cases in `_type_check_output_wrapper` for these kinds of compute functions the errors could only be caught if the wrapper was where the iteration occurred